crates/checkpoint-downloader/src/downloader.rs

```diff
@@ -379,70 +382,93 @@ impl ParallelCheckpointDownloaderInner {
     pub fn start(
-        mut self,
+        self,
         sequence_number: CheckpointSequenceNumber,
     ) -> mpsc::Receiver<CheckpointEntry> {
         let (result_tx, result_rx) = mpsc::channel(self.config.result_queue_size());
         let message_sender = self.message_sender.clone();
 
         tokio::task::spawn(async move {
-            let mut sequence_number = sequence_number;
-            let mut in_flight = BTreeSet::new();
-            let mut pending = BTreeMap::new();
-            let mut next_expected = sequence_number;
-
-            while !self.cancellation_token.is_cancelled() {
-                let num_workers = self.worker_count.load(std::sync::atomic::Ordering::Relaxed);
-                let max_in_flight_requests =
-                    num_workers * self.config.channel_config.work_queue_buffer_factor;
-                while in_flight.len() < max_in_flight_requests {
-                    if message_sender
-                        .send(WorkerMessage::Download(sequence_number))
-                        .await
-                        .is_err()
-                    {
-                        tracing::error!("failed to send job to workers; channel closed");
-                        return;
-                    }
-                    tracing_sampled::info!(
-                        "30s",
-                        "adding checkpoint to worker queue {}",
-                        sequence_number
-                    );
-                    in_flight.insert(sequence_number);
-                    sequence_number += 1;
-                }
-
-                let Some(entry) = self.checkpoint_receiver.recv().await else {
-                    break;
-                };
+            self.run_download_loop(sequence_number, result_tx, message_sender)
+                .await;
+        });
 
-                in_flight.remove(&entry.sequence_number);
+        result_rx
+    }
 
-                if entry.sequence_number != next_expected {
-                    // Store out-of-order response
-                    pending.insert(entry.sequence_number, entry);
-                    continue;
+    /// Main download loop that coordinates checkpoint fetching and result delivery.
+    async fn run_download_loop(
+        mut self,
+        sequence_number: CheckpointSequenceNumber,
+        result_tx: mpsc::Sender<CheckpointEntry>,
+        message_sender: async_channel::Sender<WorkerMessage>,
+    ) {
+        let mut sequence_number = sequence_number;
+        let mut in_flight = BTreeSet::new();
+        let mut pending = BTreeMap::new();
+        let mut next_expected = sequence_number;
+
+        while !self.cancellation_token.is_cancelled() {
+            let num_workers = self.worker_count.load(std::sync::atomic::Ordering::SeqCst);
+            let max_in_flight_requests =
+                num_workers * self.config.channel_config.work_queue_buffer_factor;
+            while !message_sender.is_full() && in_flight.len() < max_in_flight_requests {
```

Comment:
The key of the fix is adding `!message_sender.is_full()` here. I was planning to do some refactor around the cancellation token to make it more effective (as opposite to only cancel the task between checkpoints). But later found some difficulties to make it work. I preserved `run_download_loop`, but the while loop is almost identical to the old code (besides adding the metrics).